File: prettier-plugin-csharpier/src/Printer/Types/AttributeList.ts
import { Doc } from "prettier";
import { PrintMethod } from "../PrintMethod";
import { HasIdentifier, printPathIdentifier, SyntaxToken, SyntaxTreeNode } from "../SyntaxTreeNode";
import { concat, group, hardline, indent, join, softline, line, doubleHardline } from "../Builders";
import { printNameColon } from "./NameColon";
import { NameEqualsNode, printNameEquals } from "./NameEquals";

export interface AttributeListNode extends SyntaxTreeNode<"AttributeList"> {
    openBracketToken?: SyntaxToken;
    target?: AttributeTargetSpecifierNode;
    attributes: AttributeNode[];
    closeBracketToken?: SyntaxToken;
}

interface AttributeTargetSpecifierNode extends SyntaxTreeNode<"AttributeTargetSpecifier"> {
    identifier: SyntaxToken;
    colonToken?: SyntaxToken;
}

interface AttributeNode extends SyntaxTreeNode<"Attribute"> {
    name?: SyntaxTreeNode;
    argumentList?: AttributeArgumentListNode;
}

interface AttributeArgumentListNode extends SyntaxTreeNode<"AttributeArgumentList"> {
    openParenToken?: SyntaxToken;
    arguments: AttributeArgumentNode[];
    closeParenToken?: SyntaxToken;
}

interface AttributeArgumentNode extends SyntaxTreeNode<"AttributeArgument"> {
    nameEquals?: NameEqualsNode;
    nameColon?: NameColonNode;
    expression?: SyntaxTreeNode;
}

interface NameColonNode extends SyntaxTreeNode<"NameColon"> {
    name?: HasIdentifier;
    colonToken?: SyntaxToken;
}

export const printAttributeList: PrintMethod<AttributeListNode> = (path, options, print) => {
    const node = path.getValue();
    const parts: Doc[] = ["["];

    if (node.target) {
        parts.push(
            path.call(attributeTargetSpecifierPath => printPathIdentifier(attributeTargetSpecifierPath), "target"),
            ": ",
        );
    }

    const attributes = path.map(attributePath => {
        const attributeNode = attributePath.getValue();
        const name = attributePath.call(print, "name");
        if (!attributeNode.argumentList) {
            return name;
        }

        const parts: Doc[] = [name, "("];
        parts.push(
            join(
                ", ",
                attributePath.map(
                    attributeArgumentPath => {
                        const attributeArgumentNode = attributeArgumentPath.getValue();
                        return concat([
                            attributeArgumentNode.nameEquals
                                ? attributeArgumentPath.call(o => printNameEquals(o, options, print), "nameEquals")
                                : "",
                            attributeArgumentNode.nameColon
                                ? attributeArgumentPath.call(o => printNameColon(o, options, print), "nameColon")
                                : "",
                            attributeArgumentPath.call(print, "expression"),
                        ]);
                    },
                    "argumentList",
                    "arguments",
                ),
            ),
        );
        parts.push(")");
        return concat(parts);
    }, "attributes");

    parts.push(join(", ", attributes), "]");
    return concat(parts);
};

File: Src/CSharpier/SyntaxPrinter/SyntaxNodePrinters/AttributeList.cs
namespace CSharpier.SyntaxPrinter.SyntaxNodePrinters;

internal static class AttributeList
{
    public static Doc Print(AttributeListSyntax node, PrintingContext context)
    {
        if (node.Parent is BaseMethodDeclarationSyntax && CSharpierIgnore.HasIgnoreComment(node))
        {
            return CSharpierIgnore.PrintWithoutFormatting(node, context).Trim();
        }

        var docs = new List<Doc>();
        if (
            node.Parent is CompilationUnitSyntax compilationUnitSyntax
            && compilationUnitSyntax.AttributeLists.First() != node
        )
        {
            docs.Add(ExtraNewLines.Print(node));
        }

        docs.Add(Token.Print(node.OpenBracketToken, context));
        if (node.Target != null)
        {
            docs.Add(
                Token.Print(node.Target.Identifier, context),
                Token.PrintWithSuffix(node.Target.ColonToken, " ", context)
            );
        }

        var printSeparatedSyntaxList = SeparatedSyntaxList.Print(
            node.Attributes,
            (attributeNode, _) =>
            {
                var name = Node.Print(attributeNode.Name, context);
                if (attributeNode.ArgumentList == null)
                {
                    return name;
                }

                return Doc.Group(
                    name,
                    Token.Print(attributeNode.ArgumentList.OpenParenToken, context),
                    Doc.Indent(
                        Doc.SoftLine,
                        SeparatedSyntaxList.Print(
                            attributeNode.ArgumentList.Arguments,
                            (attributeArgumentNode, _) =>
                                Doc.Concat(
                                    attributeArgumentNode.NameEquals != null
                                        ? NameEquals.Print(
                                            attributeArgumentNode.NameEquals,
                                            context
                                        )
                                        : Doc.Null,
                                    attributeArgumentNode.NameColon != null
                                        ? BaseExpressionColon.Print(
                                            attributeArgumentNode.NameColon,
                                            context
                                        )
                                        : Doc.Null,
                                    Node.Print(attributeArgumentNode.Expression, context)
                                ),
                            Doc.Line,
                            context
                        )
                    ),
                    Doc.SoftLine,
                    Token.Print(attributeNode.ArgumentList.CloseParenToken, context)
                );
            },
            Doc.Line,
            context
        );

        docs.Add(
            node.Attributes.Count > 1
                ? Doc.Indent(Doc.SoftLine, printSeparatedSyntaxList)
                : printSeparatedSyntaxList
        );

        docs.Add(
            node.Attributes.Count > 1 ? Doc.SoftLine : Doc.Null,
            Token.Print(node.CloseBracketToken, context)
        );

        return Doc.Group(docs);
    }
}
